### ISSUE TITLE
Weirdness with Android Push notifications

### ISSUE THREAD

**Issue by Bobisback** [urbanairship/xamarin-component#60]
So I got the push notification working I thought. Now they seem to be working off and on with android. iOS is receiving them every time without fail. It seems to be somewhat random. If I send a test message sometimes it works sometimes not. If I send a regular message I have only had it work once out of the 10 or so times I have tried. 

The android device seems to be registering correctly. Below is the channel. In the logs it is also showing this channel as registered. The log is attached. It is also showing the opt-in as good.

```
{
      "channel_id": "98331d75-66aa-429b-9750-8c22eea6da78",
      "device_type": "android",
      "installed": true,
      "background": true,
      "push_address": "Not sure if I should post this online",
      "named_user_id": null,
      "alias": null,
      "tags": [],
      "tag_groups": {
        "ua_background_enabled": [
          "true"
        ],
        "ua_opt_in": [
          "true"
        ]
      },
      "created": "2017-03-30T15:37:46",
      "opt_in": true,
      "last_registration": "2017-03-30T15:37:46"
    }
```

If the opt-in is good this page suggests: https://docs.urbanairship.com/reference/troubleshooting/android/#not-receiving-push

"If opted-in: Your GCM/FCM push provider key (located in the Go dashboard on the Settings > Services page) may be invalid. Please review the instructions here and verify you have configured GCM correctly."

I am getting the notifications sometimes though so I am guessing it cannot be this?

[ANdroidLog.txt](https://github.com/urbanairship/xamarin-component/files/883059/ANdroidLog.txt)

I am currently using Xamarin Forms 2.3.3:193 and Urban Airship SDK 4.2.2. Any insight would be awesome.

Thanks,
Bob

P.S Notifications on the device are allowed. And I am enabled user notifications through the sdk.

**Comment by rlepinski** [comment]
@Bobisback How and where are you calling takeOff?

**Comment by Bobisback** [comment]
In  class I created

`[Application]
    public class AirshipApplication : global::Android.App.Application
    {
        public AirshipApplication(IntPtr javaReference, JniHandleOwnership transfer) : base(javaReference, transfer)
        {
        }

        public override void OnCreate()
        {
            base.OnCreate();
             
            UAirship.TakeOff(this, (UAirship airship) => {
                // Configure airship here
            });
            UAirship.Shared().PushManager.UserNotificationsEnabled = true;
        }
    }`

Now I do call the enable user notifications in that class. Could this be causing the problems? I am doing it to fast? The docs do not really say where to call that line of code.

**Comment by rlepinski** [comment]
no, that should all be fine. Technically `UAirship.Shared()` will block until airship is ready so you might want to move it into the callback. But blocking is literally a few milliseconds so its not a big deal.

**Comment by rlepinski** [comment]
I wonder if this has to do with multidex usage, what android version are you testing against?

**Comment by Bobisback** [comment]
My main phone I test on is version 6.0.1. Samsung s6 I think.

I am booting up another phone right now to test on that is version 5.0 and a note 3 I believe.

**Comment by rlepinski** [comment]
I am reading about xamarin multi dex. Usually it requires extending a multidex application, not sure how it works in xamarin. Could you try disabling multidex and proguard and setting the minSDKVersion to 23? If that still does not work than we can rule out multidex.

**Comment by Bobisback** [comment]
So this is interesting. My other phone seems to be working every time. It may be a issue with my phone. 

I will try what you ask, but I am sure it is just going to error with "java.eve exited with code 2" or something like that. if I disable mutidexing and proguard. I will let you know the results.

**Comment by Bobisback** [comment]
So I recompiled the app and it stopped working again on the other phone. Uninstalling the app manually before starting it again does seem to help though not always.

I will try the mutidexing thing next and report back. 

P.S if I go to api 23 xamarin forms will not work since it only works with mono 7.0 right? 

Edit: Oh you said min sdk. My min SDK is 14 atm. Do you want me to put my min sdk to 23?

**Comment by rlepinski** [comment]
@Bobisback So when using the standard android tools, you can set the minSDK to lollipop to avoid any multidex issues. Not sure if its the same on Xamarin or not. Lollipop+ devices have native multidex support.

**Comment by Bobisback** [comment]
ummm okie dokie I will set the minsdk to 21 and see if that helps. Basically though that means I cannot use mutidexing reliably on per api 21 devices?

**Comment by Bobisback** [comment]
Ya if I turn mutidexing off I get '"java.exe" exited with code 2.'

**Comment by rlepinski** [comment]
Well, alright. Since I am super skeptical of multidex, could you try removing the application override and instead using autopilot?

Sample autopilot:
```
using System;

using Android.Content;
using Android.Runtime;

using UrbanAirship;

namespace Sample
{
        [Android.Runtime.Preserve (AllMembers = true)]
	[Register("sample.SampleAutopilot")]
	public class SampleAutopilot : Autopilot
	{
		
		public override void OnAirshipReady(UAirship airship)
		{
			airship.PushManager.UserNotificationsEnabled = true;
		}
	}
}
```

Then register it in the assembly info:
```
[assembly: MetaData("com.urbanairship.autopilot", Value = "sample.SampleAutopilot")]
```

**Comment by Bobisback** [comment]
So mutidexing on android minsdk at api 21 5.0 still the same behavior. So the real question is will this all matter in a release build. 

Ok I will switch to autopilot. Can I still use hte properties file or should I move everything over to manually put the keys in?

**Comment by rlepinski** [comment]
It will automatically pull in the properties file

**Comment by rlepinski** [comment]
@Bobisback Are you finding the issue is per build, per app open, or when the app is in a certain state?

**Comment by Bobisback** [comment]
It looks like it is per build, doing a fresh install is the key. If I manually uninstall the app and then run the project. It seems to work. If I recompile and run the project it does not work. If I uninstall the app and recompile and run the project it works.

I am just wondering how this will affect a release build. Like if I release the build and then do a update will the push notifications stop working.

With autopilot I cannot get it to work at all. The logs never sure the user opting in but a breakpoint shows airship.PushManager.UserNotificationsEnabled = true; being called correctly. I am going to walk through the documentation to make sure I did not miss any setup steps.

**Comment by rlepinski** [comment]
@Bobisback when it stops working, can you clear the data and run the app again to see if it works?

**Comment by Bobisback** [comment]
There does seem to be some sort of error going on after creating the session.

```
03-30 13:30:51.731 I/Citizens Connect - UALib(11094): Airship taking off!
03-30 13:30:51.731 I/Citizens Connect - UALib(11094): Airship log level: 3
03-30 13:30:51.731 I/Citizens Connect - UALib(11094): UA Version: 8.3.1 / App key =  Production = false
03-30 13:30:51.771 D/Citizens Connect - UALib(11094): RichPushUser - Updating user.
03-30 13:30:51.781 D/Citizens Connect Channel ID(11094): 9cf645ba-a2be-4dcb-8d59-2302abcd4c5f
03-30 13:30:51.811 D/Citizens Connect - UALib(11094): Analytics - New session: 3995e3e3-512f-4409-9815-c1a95be2d7e7
03-30 13:30:51.821 D/Mono    (11094): Assembly Ref addref Plugin.Connectivity[0xafcbd7c0] -> Mono.Android[0xb3b2bb20]: 29
03-30 13:30:51.821 D/Mono    (11094): Assembly Ref addref Xamarin.Forms.Platform.Android[0xb3b2b2e0] -> Mono.Android[0xb3b2bb20]: 30
03-30 13:30:51.821 D/Mono    (11094): Image addref System.ObjectModel[0x9ac03940] -> System.ObjectModel.dll[0x9a625100]: 2
03-30 13:30:51.821 D/Mono    (11094): Prepared to set up assembly 'System.ObjectModel' (System.ObjectModel.dll)
03-30 13:30:51.821 D/Mono    (11094): Assembly System.ObjectModel[0x9ac03940] added to domain RootDomain, ref_count=1
03-30 13:30:51.831 D/Mono    (11094): AOT module 'System.ObjectModel.dll.so' not found: dlopen failed: library "/data/app/gov.leoncountyfl.citizensconnectmobile-1/lib/arm/libaot-System.ObjectModel.dll.so" not found
03-30 13:30:51.831 D/Mono    (11094): AOT module '/Users/builder/data/lanes/4009/3a62f1ea/source/monodroid/builds/install/mono-armv7/lib/mono/aot-cache/arm/System.ObjectModel.dll.so' not found: dlopen failed: library "/data/app/gov.leoncountyfl.citizensconnectmobile-1/lib/arm/libaot-System.ObjectModel.dll.so" not found
03-30 13:30:51.831 D/Mono    (11094): Config attempting to parse: 'System.ObjectModel.dll.config'.
03-30 13:30:51.831 D/Mono    (11094): Config attempting to parse: '/Users/builder/data/lanes/4009/3a62f1ea/source/monodroid/builds/install/mono-armv7/etc/mono/assemblies/System.ObjectModel/System.ObjectModel.config'.
03-30 13:30:51.831 D/Mono    (11094): Assembly Ref addref Xamarin.Forms.Core[0xb3b2b1c0] -> System.ObjectModel[0x9ac03940]: 2
03-30 13:30:51.831 D/Mono    (11094): Assembly Ref addref System.ObjectModel[0x9ac03940] -> System[0xaeca23a0]: 3
03-30 13:30:51.851 I/Citizens Connect - UALib(11094): Airship ready!
```

**Comment by Bobisback** [comment]
you know now that you mention mutidexing and applications I do remember a post I read about not being able to use mutidexing with a custom application.

**Comment by rlepinski** [comment]
It looks like xamarin is doing some compile time magic to make multldex work. Its usually not as magically with standard android devleopment. I would get autopilot working. Our sample we provide has a working version of autopilot if you want to check that out.

**Comment by Bobisback** [comment]
Okie dokie I will work on getting autopilot running.

In the mean time, clearing the the cache does not do anything still the same issue, but (and this is a huge butt) clearing the data does fix it...... If I clear the data it will work on the next recompile.

Yup I can confirm if I clear the 260KB of data that is stored it runs fine on the next recompile without a uninstall. I tried it twice and worked both times. Then tried it without twice and it failed both times.

Interesting. So that suggests that it is storing some information that is invalid on the next build. And for some reason that information is not being updated.

**Comment by rlepinski** [comment]
Hmm... bit weird. Is the app deleting anything from the data or cache directory?

**Comment by Bobisback** [comment]
No it should not be, I have a "DataPersistenceService" but I am not using it in this version of the app. That is an Xamarin forms implementation though so I am not sure if it uses the data or cache directory. So it should not be doing anything.

**Comment by Bobisback** [comment]
So if do not rerun the app and just try to clear the data. The push notifications do not work. Even if I restart the app.

Also I got autopilot working, it is still doing the same behavior.

**Comment by Bobisback** [comment]
At this point it seems trying to make a sample project is going to be the next option. To see if it is a problem with my app or a bug.

**Comment by Bobisback** [comment]
So at first it appeared as if the test project worked fine. After closer inspection of the properties though I notice one difference (besides mutidexing) and that was the new 7.0 feature of fast deployment. I have it uncheck in my project but it is checked by default in all new projects that target 7.1 and above. If I disable fast deployment the behavior described above comes back in the sample project.

So naturally I went to my old project and check fast deployment. I ran it on a fresh install. The first message worked, cool, but that is expected behavior. I recompiled and ran it again. Sent a message, and sure enough it worked. 

So it would appear that what ever is going on has to do with fast deployment? Idk but it does seem to be working now. Consistently. (Fingers crossed)

So on to the next issue, it is unrelated to this though and requires a bit more research for me to make sure I semi know what I am talking about. I will create a new issue for that, but to give you a heads up it is related to xamarin forms and deep linking :D

**Comment by rlepinski** [comment]
Wow. Please file bugs with Xamarin. Good job debugging everything. It sounds like it will only be an issue for development with fast deploy enabled, but you might want to try to generate few release builds and manually install them through adb to double check.

**Comment by rlepinski** [comment]
@Bobisback For deep linking, checkout out:  https://docs.urbanairship.com/guides/android-deep-linking/ https://docs.urbanairship.com/guides/ios-deep-linking/

I would override the deep link action on both platforms and do some generic forms logic instead of letting the actions do platform specific deep linking.

**Comment by Bobisback** [comment]
ya the issue is though urban airship appears to be destroying the android activity that is running xamarin forms. Because my main page is a splash screen not xamarin. But again I need to do some more research to try and track down what it is doing. I will create a new issue for it once I have more info and if/when I run into a road block.

**Comment by Bobisback** [comment]
This can be closed now.

**Comment by Arsophia** [comment]
Hi all, I have read all this thread and found that the initial described behavior is happening in my environment, deploying to Android 7.1, with fast deploy enabled, multidex on and the IDE is Visual Studio for Mac. Latest package versions for Xamarin Forms 2.4.0.282, Android Support Libraries 25.4.02 and Urban Airship Nuget 4.6.1. 

The only fix I have for now is cleaning the application data on every rebuild, however, as my App have a high security logging system, running through it every build can be a pain as you can imagine. 

Any guide will be helpful, maybe it is something about a new packet version? 

Thanks!
